### ISSUE TITLE
Add DataSubjectRegistry

### ISSUE THREAD

**Issue by madhavajay** [OpenMined/PySyft#7036]
## Description
The current data structure for `Data Subjects` is not going to scale well and is already causing some issues in the internals of the `DataSubjectLedger`.

We should revert to using unique sequential `int` values from 0 onwards for the code backed by a lookup in the database with a unique int id primary key.

`Table`: `data_subject_registry`:
`id` int64 primary key UNIQUE (int8 or bigint) 
`data_subject` str UNIQUE 255
add index for the data_subject column so we can query it
`created_at` timestamp

Private server side methods:
`get_all_ids` returns the entire table as a `np.ndarray[str]` array of strings from 0 onwards
this method needs to be paginateable so select from id 0 -> 1 million
`get_all_data_subjects` returns the entire table as a `Dict[str, int]` mapping, this method needs to be paginateable so like select from id 0 -> 1 million 
`get_all_ids_and_data_subjects` returns both as a Tuple
`register_data_subjects` takes `UnindexedDataSubjects` and does all the work to deduplicate and then insert them into the db returning `IndexedDataSubjects` object.
needs to be paginateable so like select from id 0 -> 1 million 

During `load_dataset` we need to register these data subjects by sending the unique ones to the service. Anything not already in the table (perhaps do a `get_all_data_subjects`) can be passed to the `insert` function. Exceptions for duplicate rejection should be handled.

This means the DPTensors need to have two types of `DataSubject` objects.
`IndexedDataSubjects` and `UnindexedDataSubjects`.

When users submit tensors they can have either but if they have `UnindexedDataSubjects` they need to be swapped out after Registration.

Question: @iamtrask Do we need to prevent non `DO`s from registering DataSubjects even though the internal ids are always hidden. In the case of some kind of leak of information, the ability to upload names or ids to see what they are converted to could be abused.

## Acceptance Criteria
We need to have some tests which verify that these methods work and the `UNIQUE` attributes of the `id` and `str` are maintained.

- ids should never be reused or recycled (so no row deletion)
- data_subject strings need to be unique otherwise we will get duplicates

**Comment by iamtrask** [comment]
I agree with the issue and the design recommendations. However, as it's focused on achieving scale (as opposed to a tinker-level MVP API) I think it should be pushed to 0.8, which has a scale-focused aspect.

**Comment by iamtrask** [comment]
Actually I didn't realize the full scope of this issue - that there's incorrect indexing occurring. This needs to be a 0.7 issue or it will appear that privacy budget spends very very inefficiently.

**Comment by madhavajay** [comment]
@iamtrask we can solve this faster by changing the `DataSubjectLedger` to use a Dict where the key is the DataSubject String but it will need to change later as `Dict[str,float]` wont be fast or scale like a sequential int id as indexes into a `np.ndarray[float64]`.

Do you still want us to build the DSR to ensure there is a unique sequential `int` for each `DataSubject` and then use that in the Ledger?
